### ISSUE TITLE
Errors from forum #2

### ISSUE THREAD

**Issue by SBPrime** [SBPrime/AsyncWorldEdit-Premium#12]
From forum:

this should be a quick fix (at least #1)?

-server was fully updated (new AWE 3.2.0 files) and restarted before executing this initial test.

1: SPAM: v3.2.0 awe spams chat with  https://gyazo.com/0afada4cff5f1e9b8ce59166cc66983b "out of space on AWE block queue" when it runs out of space obviously. same conditions as before     500x61x500 blocks, same config as before, updated to v9 config: http://pastebin.com/bAjNdXRy (4gb min-memory-soft limit was written to the config later, default was 3GB, what i ran the test with/on) however once it gets going, and the API starts showing the progress accurately it stops spamming it entirely. (i assume it found the 'happy medium'

2: 'KICKS': Another job was started by another user at about 70% into the above edit, (57k block paste) and that process kicked all the players from the server. so that issue is still there... I assume starting a new job initially ignores the available memory and starts it's process before considering how much memory is available - rather than checking available memory and blanching the jobs memory usage between multiple jobs. or something like that.

3: 'BLOCK LOGGING ERROR': also found multiple error messages in console related to block logging: 

```
----------repeated errors:
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] Error logging block: java.lang.ClassCastException
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] Player: leasoncre
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] World: CraftWorld{name=world}
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] Location: Location{world=CraftWorld{name=world},x=6602.0,y=27.0,z=8582.0,pitch=0.0,yaw=0.0}
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] Old: Block{ID:52, Data: 0}
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] New: Block{ID:0, Data: 0}
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] Error logging block.
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] * Exception: java.lang.ClassCastException
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] * Error message: null
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] * Stack:
[02:44:28] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
----------first error:
[02:30:56] [Server thread/INFO]: WorldEdit: leasoncre (in "world"): /set 0 - Region: (6540.0, 1.0,     8200.0) - (7040.0, 61.0, 8699.0)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] Error logging block: java.lang.ClassCastException: org.bukkit.craftbukkit.v1_8_R3.block.CraftBlockState cannot be cast to org.bukkit.block.CreatureSpawner
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] Player: leasoncre
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] World: CraftWorld{name=world}
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] Location: Location{world=CraftWorld{name=world},x=6983.0,y=39.0,z=8200.0,pitch=0.0,yaw=0.0}
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] Old: Block{ID:52, Data: 0}
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] New: Block{ID:0, Data: 0}
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] Error logging block.
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * Exception: java.lang.ClassCastException
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * Error message: org.bukkit.craftbukkit.v1_8_R3.block.CraftBlockState cannot be cast to org.bukkit.block.CreatureSpawner
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * Stack:
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.coreprotect.consumer.Queue.queueBlockBreak(Queue.java:172)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.coreprotect.consumer.Queue.queueBlockBreak(Queue.java:188)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.coreprotect.CoreProtectAPI.logRemoval(CoreProtectAPI.java:220)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.PrimeSoft.blocksHub.coreProtect.CoreProtectLogger.logBlock(CoreProtectLogger.java:128)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.PrimeSoft.blocksHub.Logic.logBlock(Logic.java:241)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.BlocksHubIntegration.logBlock(BlocksHubIntegration.java:131)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.BlocksHubIntegration.logBlock(BlocksHubIntegration.java:274)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.world.AsyncWorld.logBlock(AsyncWorld.java:1020)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.world.AsyncWorld.access$000(AsyncWorld.java:91)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.world.AsyncWorld$8.execute(AsyncWorld.java:280)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.world.AsyncWorld$8.execute(AsyncWorld.java:268)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.entries.WorldFuncEntryEx.process(WorldFuncEntryEx.java:79)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.BlockPlacer.processQueue(BlockPlacer.java:426)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.BlockPlacer.run(BlockPlacer.java:351)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.BlockPlacer.access$000(BlockPlacer.java:76)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.BlockPlacer$1.run(BlockPlacer.java:248)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.blockPlacer.BlockPlacerTask$1.run(BlockPlacerTask.java:81)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftTask.run(CraftTask.java:71)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:350)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.minecraft.server.v1_8_R3.MinecraftServer.B(MinecraftServer.java:723)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.minecraft.server.v1_8_R3.DedicatedServer.B(DedicatedServer.java:374)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.minecraft.server.v1_8_R3.MinecraftServer.A(MinecraftServer.java:654)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * net.minecraft.server.v1_8_R3.MinecraftServer.run(MinecraftServer.java:557)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] * java.lang.Thread.run(Thread.java:745)
[02:30:58] [Server thread/INFO]: [AsyncWorldEdit] ***********************************
full log: http://pastebin.com/efuYRS2b
```

**Comment by leasoncre** [comment]
Reply from discussion thread:
0. Noted. will post here.
Ad. 1. If the operation is large and the server is fast enough this might happen. Because it constantly moving from hard to soft limit. Never the less I'll see if and what I can do wit this (any suggestions?)
+create a limit on the number of times per second a message can be sent: [ie: specifically: 1x "out of space" message every 10 seconds.]
Ad. 2. I was unable to reproduce this bug.
+odd. i can make it do this on my server every time (some random delay may be involved)

Ad. 3. This does not look like AWE error. Please send me what version of CoreProtect are you using and what version of spigo.
+versions listed below. full log is posted in the link in original post. I did not have this particular issue prior to updating awe to 3.2. from 3.1.1

This server is running CraftBukkit version git-Spigot-76236cb-62a2169 (MC: 1.8.8) (Implementing API version 1.8.8-R0.1-SNAPSHOT)       [spigot-BT45_1.8.8_09.9.jar]
Version: CoreProtect v2.12.0.

**Comment by leasoncre** [comment]
apologies on not posting this here to start with.

**Comment by SBPrime** [comment]
Ad. 1. A cool-down time sounds good. 

Ad. 2. Ok since you are able to reproduce it in 100% situations please try this: enable debug mode (in awe config) then try to invoke the error and send me the server log (its going to be huge ;)), after that you can disable debug mode.

Ad. 3. I'll look into this. Is this a one time thing or is it constantly appearing? Does it impact coreprotect?

**Comment by leasoncre** [comment]
1. hymm.. well this is a pickle... i turn AWE debug on and now it won't 'crash'/kick ppl
   i tried an instance w/o debug on and successfully replicated the issue, came back, purged all jobs, re-started the operation after enabling debug and /awe reload all. (all the console spam...) and it's not 'kicking' everyone WITH debug on... and while i was looking at the log. it kicked everyone about 10 min after operation completed. 
   log: https://www.dropbox.com/s/q3vgee11tbmpb2d/latest.log?dl=0

3: it's constantly appearing during awe operations. the error is logged to console, and coreprotect does not receive the block change information

4: i found 2 sections of blocks that got missed by async when i did //set 0 operations, i don't think they occurred during the debug portion of the log though. coreprotect says the blocks were removed even though i'm staring at them.
-similar issue occurs to other users as well

**Comment by SBPrime** [comment]
Ad. 1. Still can't reproduce it... I really no idea what could be causing your error. 
- Could you try to reproduce this on a clean server (dev, test whatever)? 
- To be 100% sure its not a crash only a timeout? And the players get kicked
- Could you try to do /gc /mem while you expect the crash (from console) to see whats exactly happening with the memory.
- According to the log you have send me you ware running this from 2 groups, and you had some memory low sections
- "it kicked everyone about 10 min after operation completed." <---- this looks interesting AFTER operation completed? This could only mean memory issues with the undo queue

The CoreProtect:
- AWE is not responsible for CoreProtect integration, BlocksHub does this. Please send me the version of BlocksHub and BlocksHub-CoreProtect.
- Are you able to provide a schematic of the location where the block change causes error? (before edit from backup?)

**Comment by SBPrime** [comment]
The cooldown time is added in v3.2.1

**Comment by leasoncre** [comment]
1.replicated server (with the exception of the worlds themselves, copy-pasted plugins-folder and spigot jar. to home machine for testing, updated to async 3.2.1)
[[host machine base specs: Win7x64, 32GB ram, intel i7-4770@3.4GHz]]
++can prob skip to the 'clean' test section....++

-----HDD test: (where i first had the files, later moved to SSD with no relative change to operations aside from speed and much less frequent 'pauses' to server operation (as expected from an SSD))
-it does look like the server is operating under the 2GB free memory limit. noticed it seems to 'seize up' for a moment when memory appears to run out (hard to tell with only console). 
-i would assume by this that users logged into a remote server (as opposed to on one's own host machine) are getting 'kicked' due to internet timeout protocol, my machine may be slower due to HDD instead of SSD.
-and finally replicated, got kicked (twice), but server still running. canceling jobs, and restarting on an SSD. at this point console is rather slow to respond to commands (several seconds before what i type even shows up) -coreprotect never finished it's shutdown sequence. killed process.
-'server low on memory' message still spams the first time it occurs, to a lesser degree though (3-5 times instead of ~15)
-odd thing is log only shows the vip group debug.... sarina does not have this.

----RETESTED on SSD, closed all non-essential programs on host machine (expect MC x2, the server console, some fileapps, and firefox for writing this. test was certainly alot smoother):

-restarted //set 0 on 2 500x61x500 areas, (i'd assume sarina had vip 'cuz she was opped, but still seems odd. as this time she isn't opped)
-not spamming the memory low now either. suspect previous test was not accurate/updated properly.
-results are the same as i have been haveing since i started this ticket. got kicked again
-spammed console with /gc command on this test, hope it helps

------'CLEAN' TEST ON SSD: (removed all plugins excpect CoreProtect,BlocksHub,Async,Actionbar,WorldEdit, Essentials/Groupmanager. kept configs)
-replicated the 'kicking' much faster on the first attempt. logged back on and restarted canceled operations (sarina's //set 0) and seems to be harder to replicate now....
-memory doesn't clear much after sustained operations. //clearhistory doesn't help. -wow... derp... it would help if i had generated the chunk first. >_>
-'out of space' spams ~10times still, generally only during thew beginning several minutes of the operation
-spammed /gc command here too (except when que locked)
-unable to replicate kicking users a second time with this configuration. memory still hits around 500mb free and 'freezez'
-it also looks like one job or the other will get 'stuck' and not que more blocks when memory frees up 
-physics freeze is set to true, but doesn't always freeze every block with physics. probably related to memory issue. https://gyazo.com/31fac1d61e86a3b53c2880c5202e084d
-took forever, but got a second disconnect to occur. has to be low memory issue.

3.blockshub log issue: (i keep everything updated as developers put it out) blocks hub v1.2, blockshub-coreprotect v1.2.1, coreprotect 2.12.0, async 3.2.1
- i don't understand what help a schematic would have. and/or what edit from backup means.

5.update/clarified/new: when useing the //undo or //undo <#> commands,  1/2 if not more of the undo does not take place, only some of the blocks are undone. try replicating via: simple: //undo, //undo <#> | complex: various number of operations involving //set, //undo, //undo <#>, //redo commands

6.vip group doesn't get too many notifications vs default (has always been like this, but not of great concern to me): https://gyazo.com/a93d4dae7aa6165b476b0c1a36889848

7.can you implement a task to re-connect the API to the user after a user logs off and logs back in, (and has a continuing operation). currently if i get kicked/disconnected the ActionBarAPI and/or AWE won't send existing job data to me after i reconnect.

LOG FILES: HDD test: https://www.dropbox.com/s/uip0s1yjf2mediv/HDD%20test.log?dl=0  SSD test: https://www.dropbox.com/s/i7vul23u63g9glc/SSD%20test.log?dl=0 'CLEAN' SSD test: https://www.dropbox.com/s/xccaoxad0ybtp8w/CLEAN%20SSD%20test.log?dl=0
Async config: https://www.dropbox.com/s/epc468e8k9tqqud/Async%20config.yml?dl=0
Blocks hub Config: https://www.dropbox.com/s/6fh1309m1khh6yr/BlocksHub%20config.yml?dl=0
CoreProtect config (if this is even needed): https://www.dropbox.com/s/3yc9vzdguj9kv5p/CoreProtect%20config.yml?dl=0

i hope i'm not killing you with all the bug reports.... and thank you for your patience :)

**Comment by SBPrime** [comment]
1. COOLDOWN

First thing: it looks like the config did not update to the newest one. If this did not update the cooldown could work incorrectly. How does the cooldown work: 
AWE sends a message and records the time. When next message gets send AWE checks if the message is the same and the time is less then the cooldown time the message is skipped. If the message is different the message is always shown. If you have problems with the cooldown please start a new issue (its much easier to handle the issues like that :P). If it works like I explained and you think it should work differently also let me know in a new issue.
1. PROGRESS DISPLAY RESTORE
   This is implemented in AWE and should work. According to your configuration you have cleanup on logout enabled. Could you clarify that this happens for group default or vip? I started an issue for this please write additional details there (#18).
2. CORE PROTECT
   I asked for the schematic because it looks like BlocksHub is unable to record a certain block type. Thats why I wanted to know what block exactly causes the error.
3. ".update/clarified/new" ummm... say what? I dont understand what do you mean could you rephrase that? If its a new bug start a new issue.

If you have ANY NEW errors/requests please start a new issue :P

Regarding the root issue: I'll get back to you after looking into the provided logs in more detail.

**Comment by SBPrime** [comment]
I think that I found the issue: You are experiencing lag spikes

```
(some examples)
[15:11:51] [Server thread/INFO]: [AsyncWorldEdit] [BP RUN] Group: AWE.Groups.vip    Blocks: 3769    Time: 25611 Demanding: N
[15:12:09] [Server thread/INFO]: [AsyncWorldEdit] [BP RUN] Group: AWE.Groups.vip    Blocks: 3928    Time: 17769 Demanding: N    
[15:12:45] [Server thread/INFO]: [AsyncWorldEdit] [BP RUN] Group: AWE.Groups.vip    Blocks: 3349    Time: 20491 Demanding: N
```

Could you try to reproduce this on a test server WITHOUT blockshub, and with chunk freeze disabled (3 tests: 1 without blockshub, 1 without physics freeze, 1 without both)

The SSD test: it looks like you run out of memory, did the user get kicked (timedout)?

If you think this might help you can contact me directly on skype (SBPrime).

**Comment by SBPrime** [comment]
I forgot to mention that the purge command is on my todo list (its going to be called from other thread then the console).

**Comment by SBPrime** [comment]
Changed how groups work, instead of separate time and block counters groups use a shared counters.
Fixed bukkit scheduler task memory leak
Added better memory handling
Added force GC when memory is running low
Added disable undo option for groups.

If I missed something please start a new issue. The core issue of this ticket is: server kicking players when running multiple tasks.

All those features are going to be included in the next release.
